### ISSUE TITLE
Reduce LSP Payload sizes / traffic

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#53501]
Currently the LSP protocol traffic includes more than it needs to (both unnecessary messages and large messages). When running locally the payload size is not a particularly significant factor in performance (because it goes between processes on the same machine) but for browser-based editors like IDX/Codespaces and remote workspaces this can significantly hurt performance.

This issue is for tracking some specifics that we may be able to improve. Some are trivial and some are more complex and may require their own issues (or upstream VS Code/LSP issues).

- Analysis
  - [ ] Initial analysis sends an unnecessary `publishDiagnostics` with an empty set for every analyzed file
  - [ ] We send several messages to for progress of analysis start/stop on each keypress.. maybe these should be debounced?
- Completion
  - [ ] Completion includes documentation for all already-imported items (which for Flutter is significant) instead of lazily using `/resolve`
  - Not-yet-imported completion items have a large `data` payload to support `/resolve`..
    - [ ] Element IDs are long (the first 2 segments might always be the same, the last might always match label?)
    - [ ] The full completion file path is repeated in all not-yet-imported items (`data.file`) but is the same
  - [ ] In VS Code we default to a large `maxCompletionItems` instead of truncating with `isIncomplete=true`
    To fix this we might need to ensure exact matches are always included, regardless of ranking (perhaps truncating based on fuzzy score and not rank?)
  - [ ] `labelDetails` is sometimes included as an empty object unnecessarily
  - [ ] `sortText` is `(10000000 - $rank)` resulting in longer strings than may be necessary
  - [ ] `additionalTextEdits` is sometimes included as an empty list unnecessarily (snippets?)
  - [ ] `documentation` is allowed to be `string` which could save including an object to mark it as markdown on each item (may be unnecessary if we push all docs to `/resolve`)
  - [ ] Snippets duplicate range that's already in `itemDefaults`

(@bwilkerson @srawlins fyi)

**Comment by DanTup** [comment]
Prompted by https://github.com/dart-lang/sdk/issues/55790, I had a look back to remind myself why this hadn't been done (and I'm noting it here so I don't have to re-discover this next time it comes up 🙂):

> Completion includes documentation for all already-imported items (which for Flutter is significant) instead of lazily using `/resolve`

We're already doing this for not-yet-imported completions so it seems like it should be simple. I started on the change and realised that in order to do this we need to round-trip _both_ the "element location" reference _and_ the "file we will insert the completion into" in the payload (we use `file` to get a session to get back to the `Element` to get the docs). This adds a fair bit to the payload of every item:

```js
"data": {
	"file": "file:///C:/Dev/Google/dart-sdk/sdk/pkg/analysis_server/lib/src/lsp/handlers/handler_completion.dart",
	"ref": "package:collection;package:collection;LinkedHashMap"
},
```

This is probably still smaller than a lot of the Flutter docs, so it might still be worthwhile - however to optimise this, we'd probably only want to do it "if the docs we would otherwise inline are greater than around 200 characters".

A better solution would be if we could eliminate duplicating "file" in every single element, which is something being discussed in https://github.com/microsoft/language-server-protocol/issues/1802 (but it's not clear if an agreement has been made yet), and then the threshold would be more like 80 characters (at which point maybe we could just include it if there is _any_ docs for consistency).

I will ping on https://github.com/microsoft/language-server-protocol/issues/1802 again, but if it's unlikely any progress will be made on that soon, then we could just implement this with a threshold of around 200chars.

(for my own future reference, the work I started on this is on a local Git branch named `resolve-completion-documentation`)

**Comment by bwilkerson** [comment]
> This adds a fair bit to the payload of every item ...

I don't know whether it's a good idea, but we could potentially add a "completion resolution cache" to the server that would map a unique suggestion id to the data we need in order to resolve the suggestion. The id could then be fairly small (a hex encoded integer, for example), minimizing the impact on the payload. The cache could be flushed whenever a file is edited, saved, or a new completion request is received.

Caching can be hard to get right, but if the payload size is a big enough issue it might be worthwhile.

**Comment by DanTup** [comment]
Yeah, we'd discussed that in the past (actually, I think we had such a cache in the past!) and it came up in https://github.com/microsoft/language-server-protocol/issues/1802, but LSP doesn't make a lot of guarantees around when a client might be allowed to call `resolve`.

Probably it would be fine though - especially if we made sure we fail gracefully (eg., at worse you might call `resolve` for an old completion item and just get back the same item without any additional docs/etc.).

I did ping on the LSP issue again - if there is no response (or appetite to progress this soon), I may look at something along these lines. There are some potentially big savings here.

**Comment by DanTup** [comment]
I've implemented some more of the items listed above as part of LSP v3.18 (it requires the client support to get the benefit). I've also crossed a few out that I no longer think are worthwhile or are not possible (because I made bad assumptions).

The only item left is truncating the completion list shorter and using `isIncomplete=true`, however the behaviour of that in VS Code is much worse than without:

- https://github.com/microsoft/vscode/issues/147830
- https://github.com/microsoft/vscode/issues/168329

So for now, I think there's nothing more to do. We can keep this issue open in the hope the VS Code issues get resolved, or we could consider this done as best as we can for now.

**Comment by bwilkerson** [comment]
I'm fine with either choice. It it would be helpful for you, then keep it open. It's you'd rather close it until there's more work to be done, then that's what we should do.

**Comment by DanTup** [comment]
Let's close this, since I do still have https://github.com/Dart-Code/Dart-Code/issues/3894 open in the Dart-Code repo about that same issue (but it's specifically about `isIncomplete` and labelled as blocked on VS Code). Thanks!
